src/resources/ble_esp32.cc

```diff
@@ -697,6 +918,112 @@ PRIMITIVE(request_attribute) {
   return process->program()->null_object();
 }
 
+/*
+ *
+ * Primitives for BLE server
+ *
+ */
+PRIMITIVE(server_config_init) {
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (!proxy) ALLOCATION_FAILED;
+
+  // Mark usage. When the group is unregistered, the usage is automatically
+  // decremented, but if group allocation fails, we manually call unuse().
+  BLEEventSource* ble = BLEEventSource::instance();
+  if (!ble->use()) {
+    MALLOC_FAILED;
+  }
+
+  BLEServerConfigGroup* group = _new BLEServerConfigGroup(process, ble);
+  if (!group) MALLOC_FAILED;
+
+  proxy->set_external_address(group);
+  return proxy;
+}
+
+PRIMITIVE(add_server_service) {
+  ARGS(BLEServerConfigGroup, group, Blob, uuid_blob);
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (!proxy) ALLOCATION_FAILED;
+
+  ble_uuid_any_t uuid = uuid_from_blob(uuid_blob);
```

Comment:
This doesn't allocate in the malloc heap, right?

Reply:
I do not understand the comment.